### ISSUE TITLE
the correct way to get_PSR_freq

### ISSUE THREAD

**Issue by luojing1211** [nanograv/PINT#713]
Currently get_PSR_freq(modelF0=True) only returns F0  in the model, however, should we get the F0 at TOA? F0(TOA) = F0 + dt * F1 +.... 
Secondly, this check seems duplicated with the Spindown component's validation. 

https://github.com/nanograv/PINT/blob/6fbbd924ff6b9a1c455ffd1351286a75442fe130/src/pint/residuals.py#L135-L156

**Comment by paulray** [comment]
When we need precise frequency at a TOA (e.g. in zima.py), I use this:
```
F_local = m.d_phase_d_toa(ts)
```
I seem to recall testing that in residuals but it makes it really slow. It is an outstanding problem that F0 is used for frequency in some places like this. It is often a sufficient approximation, but not always. I timed a pulsar that changed in period by a factor of two in a couple years.

**Comment by paulray** [comment]
See this discussion: https://github.com/nanograv/PINT/issues/415

**Comment by luojing1211** [comment]
Yeah, I remember that. Should we at least provide a Taylor expansion version of F0 at TOA? Plus, the check should use the spindown validation.

**Comment by paulray** [comment]
Sure, those changes would be fine with me.

**Comment by scottransom** [comment]
Could he have it as an option to use the full `F_local = m.d_phase_d_toa(ts)` but use the quick normal version by default?

**Comment by paulray** [comment]
That sounds fine.  For option 3, it will need to be passed a TOAs object as an argument.
I think the functionality of getting the pulsar freq is used elsewhere in PINT, so we should make them all use this one function, with its three options depending on their needs for speed and/or accuracy.

**Comment by luojing1211** [comment]
Yes, for option 2, one will need a timestamp at least.
-Jing

**Comment by paulray** [comment]
Copying this from a different discussion with @aarchiba :

> Ah, so there are four possible ways to handle phase -> time conversion:
> 
> - Use the full delay calculation (at what time? before or after delay corrections?)
> - Use the Taylor series but don't worry about delay/timescales
> - Use F0 from the model
> - Use some fiducial F0 that is not a fit parameter

We should make a PR to address this, after #861 is merged. For the full delay calculation, isn't the right time just the topocentric time of the TOA?  You want the observed apparent frequency of the pulsar to convert a phase offset to a time offset.

**Comment by abhisrkckl** [comment]
Fixed in 
https://github.com/nanograv/PINT/blob/master/src/pint/residuals.py#L246

**Comment by abhisrkckl** [comment]
New improvements to this are in #1334
